plugins/scale/wayfire/plugins/scale-transform.hpp

```diff
@@ -0,0 +1,282 @@
+            left   = std::max(left, other.left);
+            bottom = std::max(bottom, other.bottom);
+            right  = std::max(right, other.right);
+        }
+    };
+
+    uint32_t get_z_order() override
+    {
+        /* ensure that scale comes after fire */
+        return wf::TRANSFORMER_HIGHLEVEL + 2;
+    }
+
+    /**
+     * Effect hook called in the following circumstances:
+     *  (1) scale's animation is running
+     *  (2) another transformer changed the view's size, so the overlay and
+     * padding might need to be updated
+     *
+     * The second case can happen just during rendering the view (from
+     * view::render_transformed()), so it should not call damage.
+     *
+     * @return Whether the overlay changed. In this case damage will be
+     * scheduled for the view.
+     *
+     * Damage will also be scheduled if the padding's size changed.
+     */
+    using pre_hook_t = std::function<bool (void)>;
+
+    /**
+     * Render hook for drawing on top of the surface after the transform, during
+     * rendering the view.
+     *
+     * @param fb     The framebuffer to render to.
+     * @param damage Damaged region to render.
+     */
+    using render_hook_t = std::function<void (const wf::framebuffer_t& fb,
+        const wf::region_t& damage)>;
+
+    /**
+     * Overlays that can be added to this transformer. Hooks are called
+     * similarly to render-manager.
+     */
+    struct overlay_t : public wf::custom_data_t
+    {
+        /* Pre hook; called just before rendering, can adjust padding. Return
+         * value indicates if damage should be scheduled for the view. */
+        pre_hook_t pre_hook;
+        /* Render hook; called during rendering, after this transform has been
+         * applied to the view. This can only render to the view's texture. */
+        render_hook_t render_hook;
+        /* Extra padding around the transformed view required by this overlay.
+         * This is added to the view's bounding box */
+        padding_t view_padding;
+        /* Extra padding taken to be taken into consideration by scale's layout.
+         * This can differ from view_padding e.g. if this overlay is rendering
+         * directly to the end framebuffer */
+        padding_t scale_padding;
+    };
+
+    /* render the transformed view and then add all overlays */
+    void render_with_damage(wf::texture_t src_tex, wlr_box src_box,
+        const wf::region_t& damage, const wf::framebuffer_t& target_fb) override
+    {
+        /* render the transformed view first */
+        view_transformer_t::render_with_damage(src_tex, src_box, damage, target_fb);
+
+        /* call all overlays */
+        for (auto& p : overlays)
+        {
+            auto& ol = *(p.second);
+            if (ol.render_hook)
+            {
+                ol.render_hook(target_fb, damage);
+            }
+        }
+    }
+
+    /* call pre-render hooks; the parameter indicates if it is possible to
```

Comment:
I find this a bit confusing. It both damages and returns whether damaging is necessary? Looking at the current usages, it seems we could directly damage (if damaging is possible at all).

Reply:
This is a small optimization to avoid calling `damage()` twice, since this function is called from [here](https://github.com/dkondor/wayfire/blob/8d68476114371820ccb58a301bc70079476e7d23/plugins/scale/scale.cpp#L694) where damage might be necessary anyway.

I could either remove this if it is not worth the code complexity (since the runtime saving will be small), or use an alternate implementation with a second bool parameter that is set to true if the caller needs to damage the view. This would allow to only damage once even if the overlays' total size has shrunk.

Reply:
I've changed the logic here, and made the version with the `can_damage` parameter protected since it is only used when this is called from `get_bounding_box()`